collection/Permutations.java

```diff
@@ -0,0 +1,91 @@
+package com.vaticle.typedb.common.collection;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.vaticle.typedb.common.collection.Collections.list;
+import static java.util.Collections.emptyIterator;
+import static java.util.Collections.singletonList;
+
+public class Permutations {
+
+    /**
+     * We implement the C++ STL next_permutation method of lazily generating permutations
+     */
+    public static <T> Iterator<List<T>> permutations(Set<T> items) {
+        if (items.size() == 0) return emptyIterator();
+        else if (items.size() == 1) return singletonList(list(items.iterator().next())).iterator();
+
+        // assign a comparable ordering over the items
+        Map<Integer, T> mapping = new HashMap<>();
+        List<T> sortedItems = items.stream().sorted(Comparator.comparing(Object::toString)).collect(Collectors.toList());
+        int[] itemKeys = new int[sortedItems.size()];
+        for (int i = 0; i < sortedItems.size(); i++) {
+            mapping.put(i, sortedItems.get(i));
+            itemKeys[i] = i;
+        }
+
+        return new Iterator<List<T>>() {
```

Comment:
are we sure this should be an anonymous class? seems pretty involved and significant?